front_end/content.py

```diff
@@ -1103,12 +1104,26 @@ def get_exercise_submissions(self, course_id, assignment_id, exercise_id):
                       FROM course_registrations
                       WHERE course_id = ?
                    )
-                 GROUP BY s.user_id
-                 ORDER BY u.family_name, u.given_name'''
+                ORDER BY u.family_name, u.given_name'''
 
+        # add submissions to a dict
+        user_dict = {}
         for submission in self.fetchall(sql, (course_id, assignment_id, exercise_id, course_id,)):
-            submission_info = {"user_id": submission["user_id"], "name": submission["name"], "code": submission["code"], "score": submission["score"], "passed": submission["passed"]}
-            exercise_submissions.append([submission["user_id"], submission_info])
+            partner_name = self.get_user_info(submission["partner_id"])["name"] if submission["partner_id"] else None
+            submission_info = {"user_id": submission["user_id"], "name": submission["name"], "code": submission["code"], "score": submission["score"], "passed": submission["passed"], "partner_name": partner_name}
+            if submission["user_id"] in user_dict:
+                user_dict[submission["user_id"]].append(submission_info)
+            else:
+                user_dict[submission["user_id"]] = [submission_info]
+
+        # retrieve most recent submission for each student and determine whether they have passed the exercise previously
+        exercise_submissions = {}
```

Comment:
This will work, but it is less efficient because we have to pull all of the submissions to know whether each student passed. We can make it more efficient by modifying the SQL statement to use a GROUP BY. However, this SQL statement is complex, so I can work on it. Please create a detailed card in Trello that points me to this part of the code so I can work on that when I have time.

Reply:
That is a great point. I just wrote the card! I'm excited to see how you solve this, I was wondering how to do that